File: easy/arrays/python/two_sum_1.py
def twoSum(nums: list[int], target: int):
    """
    :type nums: List[int]
    :type target: int
    :rtype: List[int]
    
    difference + numberInArray = target

    difference = target - numberInArray
   
    if commonDifference not found in array
        
        store unique differen and index
        [difference] = index
    
    else
        
        otherwise return indexes found
        result = [[difference], currentIndex]

    is the difference exists if not add in the dictionary

    """
    differences = {}
    result = []
    for index, current_number in enumerate(nums):
        current_difference = target - current_number
        if current_difference in differences:

            result.append([differences[current_difference], index])
            
            return result
        # if no common differences then store current number with their index
        # this is same as eliminating possible number
        differences[current_number] = index 
    return result 

result = twoSum([2,7,11,15], 9)

print(result)
    

File: easy/arrays/cpp/two_sum_1.cpp
#include <stdio.h>
#include <vector>
#include <iostream>
#include <map>

using namespace std;
vector<int> twoSum(vector<int>& nums, int target){
	int numberInArray = 0;
	int difference = 0;
	int currentDifference = 0;
	vector<int> result;
	map<int, int> differences;
	for(size_t index = 0; index < nums.size(); index++)
	{
		int currentNumber = nums[index];

		int currentDifference = target - currentNumber;

		if (differences.count(currentDifference) > 0)
		{
			// store answer
			result.push_back( differences[ currentDifference ] );
			result.push_back( index );
			return result; 
		}

		// insert unique value and index
		differences[ currentNumber ] =  index;
	}


	    
	return {0, 0};
}

int main(void)
{
	vector<int> nums = {2,7,11,15};
	vector<int> result = twoSum(nums, 9);
	for(size_t index = 0; index < result.size(); index++)
	{
		printf("%d \n", result[index]);
	}
	return 0;
}
    
